### PR TITLE
[SPARK-10605][SQL] Create native collect_list/collect_set aggregates

### PR DESCRIPTION
**Pull request by hvanhovell**
## What changes were proposed in this pull request?

We currently use the Hive implementations for the collect_list/collect_set aggregate functions. This has a few major drawbacks: the use of HiveUDAF (which has quite a bit of overhead) and the lack of support for struct datatypes. This PR adds native implementation of these functions to Spark.

The size of the collected list/set may vary, this means we cannot use the fast, Tungsten, aggregation path to perform the aggregation, and that we fallback to the slower sort based path. Another big issue with these operators is that when the size of the collected list/set grows too large, we can start experiencing large GC pauzes and OOMEs.

This `collect*` aggregates implemented in this PR rely on the sort based aggregate path for correctness. They maintain their own internal buffer which holds the rows for one group at a time. The sortbased aggregation path is triggered by disabling `partialAggregation` for these aggregates (which is kinda funny); this technique is also employed in `org.apache.spark.sql.hiveHiveUDAFFunction`.

I have done some performance testing:

``` scala
import org.apache.spark.sql.{Dataset, Row}

sql("create function collect_list2 as 'org.apache.hadoop.hive.ql.udf.generic.GenericUDAFCollectList'")

val df = range(0, 10000000).select($"id", (rand(213123L) * 100000).cast("int").as("grp"))
df.select(countDistinct($"grp")).show

def benchmark(name: String, plan: Dataset[Row], maxItr: Int = 5): Unit = {
   // Do not measure planning.
   plan1.queryExecution.executedPlan

   // Execute the plan a number of times and average the result.
   val start = System.nanoTime
   var i = 0
   while (i < maxItr) {
     plan.rdd.foreach(row => Unit)
     i += 1
   }
   val time = (System.nanoTime - start) / (maxItr * 1000000L)
   println(s"[$name] $maxItr iterations completed in an average time of $time ms.")
}

val plan1 = df.groupBy($"grp").agg(collect_list($"id"))
val plan2 = df.groupBy($"grp").agg(callUDF("collect_list2", $"id"))

benchmark("Spark collect_list", plan1)
...
> [Spark collect_list] 5 iterations completed in an average time of 3371 ms.

benchmark("Hive collect_list", plan2)
...
> [Hive collect_list] 5 iterations completed in an average time of 9109 ms.
```

Performance is improved by a factor 2-3.
## How was this patch tested?

Added tests to `DataFrameAggregateSuite`.

### CODE REVIEW

**Comment by hvanhovell** [comment]
cc @mccheah I rebased and simplified my PR, could you take a look. This is an alternative to https://github.com/apache/spark/pull/11688.

**Comment by maver1ck** [comment]
Hi,
What about this patch ?

**Comment by hvanhovell** [comment]
I am not sure if there is enough support to add this to Spark. The thing is that this is a potential source OOME's and that it banks on the specific behavior of the `SortBasedAggregate` code path; this is however the same for Hive's `collect*` functions and this PR **is** an improvement over its `Hive` counterparts. Do you have a very pressing usecase for this?

A more fruitfull approach would be to implement a dedicated operator for this. This would eliminate the reliance on the `SortBasedAggregate`, but it would still be capable of causing OOME's (we could spill the elements to disk, but the resulting row still has to fit into main memory).

@rxin what is your take on this?

**Comment by rxin** [comment]
It's probably a good idea to have a native implementation so we don't need to fall back to Hive, and then improve on that in the future to have a sort-based approach.

**Comment by maver1ck** [comment]
There is one more thing.
We observed that collect_list doesn't work in Spark 2.0
https://issues.apache.org/jira/browse/SPARK-15293

**Comment by rxin** [comment]
I'm going to merge this in master/2.0. Thanks!

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 1]
```diff
@@ -219,20 +219,4 @@ private[sql] class HiveSessionCatalog(
...
-      createTempFunction(functionName, info, builder, ignoreIfExists = false)
-  }
-}
-
-private[sql] object HiveSessionCatalog {
-  // This is the list of Hive's built-in functions that are commonly used and we want to
-  // pre-load when we create the FunctionRegistry.
-  val preloadedHiveBuiltinFunctions =
-    ("collect_set", classOf[org.apache.hadoop.hive.ql.udf.generic.GenericUDAFCollectSet]) ::
-    ("collect_list", classOf[org.apache.hadoop.hive.ql.udf.generic.GenericUDAFCollectList]) :: Nil
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: nice

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  override def update(b: MutableRow, input: InternalRow): Unit = {
+    buffer += child.eval(input)
+  }
+
+  override def merge(buffer: MutableRow, input: InternalRow): Unit = {
+    sys.error("Collect cannot be used in partial aggregations.")
+  }
+
+  override def eval(input: InternalRow): Any = {
+    new GenericArrayData(buffer.toArray)
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala]: oh, I see. We have an internal buffer and only generate this array data when we call eval.

**Comment by yhuai** [comment]
LGTM

**Pull request closed without merging**